src/components/Navbar/Navbar.tsx

```diff
@@ -0,0 +1,177 @@
+import React, {
+  useCallback,
+  useContext,
+  useEffect,
+  useRef,
+  useState,
+} from "react";
+import { IconDefinition } from "@fortawesome/free-regular-svg-icons";
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import { ThemeContext, ThemeProvider } from "../../theme";
+import { isNavbarItem } from "./Navbar.utils";
+import { isDev } from "../../utils";
+import useDotAnimation from "./Navbar.hooks";
+import style from "../../variables.module.scss";
+import styles from "./navbar.module.scss";
+import classNames from "classnames/bind";
+
+const cx = classNames.bind(styles);
+
+const NavbarContext = React.createContext<{
+  items?: React.RefObject<HTMLLIElement>[];
+  activeId?: number;
+  setActiveId?: (id?: number) => void;
+  registerItem?: (item: React.RefObject<HTMLLIElement>) => void;
+  unRegisterItem?: (item: React.RefObject<HTMLLIElement>) => void;
+}>({});
+
+type NavbarItemProps<T extends string> = {
+  as: "a" | "button";
```

Comment:
bug: This is fragile. Imagine the dev only passes the `a` without `href`. The DOM is invalid and the link won't work, but the app won't throw any error either. In other words, both `a` and `href` are required together, otherwise, the app should break somehow. How would you solve that?

Reply:
added discriminated unions to handle it at type level

Reply:
It's better but have you really tested? Because I can still pass invalid states. For example:
- If I pass `as="a"` it doesn't complain that the "href" is missing.
- If I pass `as="button" href="foo"` it doesn't complain that "href" is not allowed.

How would you solve that?

Give it a try first. Then, you can expand the solution:

<details>

```ts
// Bonus: Reuse the common props to avoid WET
type NavbarItemCommonProps<T extends string> = {
  icon: IconDefinition;
  title: T;
  onSelect: ({ title, id }: { title: T; id?: number }) => void;
  id?: number;
  iconStyles?: React.CSSProperties;
};

type NavbarItemProps<T extends string> =
  | ({
      as: "a";
      href: string;
    } & NavbarItemCommonProps<T>)
  | ({
      as: "button";
      href?: never; // "Discriminated union" - disallow a given param.
    } & NavbarItemCommonProps<T>);
```

</details>

Reply:
<img width="741" alt="image" src="https://github.com/alexodan/navbar-challenge/assets/35075089/425dc923-9182-45a4-9332-87c7ed75534c">
<img width="879" alt="image" src="https://github.com/alexodan/navbar-challenge/assets/35075089/2910539a-510d-42e8-adca-6ec65644141c">
the errors themselves are super confusing but I get them when I try to pass button and href or "a" without href 🤔 I don't think is ideal tho!

Reply:
I think I might be getting the error but for a different reason, since href should be never if as is button type